### ISSUE TITLE
error detection - feature request?

### ISSUE THREAD

**Issue by valentin-krasontovitsch** [getsentry/sentry-go#91]
Hi there! I've finally come around to testing the new sentry SDK - looks really nice, but I'm having one problem. I don't seem to be able to figure out how I can capture the following error:

We're in a corporate setting with self-signed certificate for the sentry server. Now I know that you can set the TLS options to not enfore certificate verification, but I prefer to just use http instead.

Now when I deploy an app with sentry and forget to remove the `s` from the `https` in the sentry dsn, and I use something like

```
  sentry.CaptureException(err)
  sentrySucceeded := sentry.Flush(5 * time.Second)
```

I expect that I could use `sentrySucceeded` to see if my message was sent successfully.

When turning on debug logging for sentry, I can see that it logs that the message failed to send because of a self-signed not trusted certificate, but the message gets processed and I guess that's why Flush returns `true`.

My question is: is there any way to capture / check that error? Or is it sentry philosophy that one cannot catch such an error?

I recall discussing this a little bit in the slack channel, a while ago, but maybe it was mostly about handling sync vs. async capturing... anywho, I cannot find anything there (on slack) anymore, unfortunately... maybe you could just be kind enough to remind me what was agreed upon : )

**Comment by valentin-krasontovitsch** [comment]
Hmm pondering the code some more, I seem to remember maybe that I was told that one could just reimplement the transport interface with one's own implementation that might take care of any errors...

i guess i'm too lazy to do that 😅

so I guess the question is: is there a nice way of using the error [produced here](https://github.com/getsentry/sentry-go/blob/eaa7a3e1d1d4ead9d23d4d40c0f72ce439696ad3/transport.go#L226) that might be beneficial to other, without breaking anything?

**Comment by rhcarvalho** [comment]
Hi @valentin-krasontovitsch! Thanks for trying out the new SDK 😄 

The return value from [`sentry.Flush`](https://godoc.org/github.com/getsentry/sentry-go#Flush) indicates whether the timeout was reached, and not whether the sending of events was successful.

In most cases you don't need to call `Flush`. One time that is it useful is when your program is shutting down, and you want to wait until pending events are sent to Sentry.

> is there a nice way of using the error [produced here](https://github.com/getsentry/sentry-go/blob/eaa7a3e1d1d4ead9d23d4d40c0f72ce439696ad3/transport.go#L226) that might be beneficial to other, without breaking anything?

When `Flush` is called, the queue may have zero, one or more events that may get sent. It is not clear what error value `Flush` should return, if any.

The recommendation for testing that the DSN is setup correctly is to send test events.

Would you need any help with that?

**Comment by valentin-krasontovitsch** [comment]
Hey @rhcarvalho thanks for your response!

So I understood Flush almost correctly.

Besides that though, I wouldn't insist on Flush returning the error I speak of - rather, for instance, one could use the existing `HTTPSyncTransport` and extend that particular transport implementation by adding a method called, e.g., `SendEventAndConfirm`, which does exactly the same thing as `SendEvent`, besides that it not only logs [the error occuring here](https://github.com/getsentry/sentry-go/blob/4c21269010222e13a2278e9249ae4102d428c7ae/transport.go#L322), but also returns it.

This error of course would have to be returned by the client method that called the transport method...

I could prepare a pull request, by extending the Client API, as I did on the previous SDK. But I would first like to know if this is at all in the interest of the community.

To explain our use case: For us, errors in apps don't occur that often, but when they occur, it's really important that we know about them. If sentry doesn't get a message about the exception though, then we don't (get notified by email), and that's so bad, that we would like to escalate those cases. We want our app to die if it doesn't manage to report to sentry - because that's another way of triggering a message to us (the developers)...

**Comment by rhcarvalho** [comment]
@valentin-krasontovitsch thanks for explaining your use case, really appreciated.

tldr; I've discussed your use case with the team and we'll brainstorm if there's something that could be done across all SDKs in cases like that.

For what it's worth, the new SDKs follow a unified concept https://docs.sentry.io/development/sdk-dev/.

---

> We want our app to die if it doesn't manage to report to sentry

With the current sentry-go, the answer is indeed to use a custom transport implementation.

Another suggestion is have your program send a message to Sentry on start up and check them on the web.

Example code: https://gist.github.com/rhcarvalho/c398236854c3dce3532e3eb428701fef

**Comment by valentin-krasontovitsch** [comment]
Thanks again for your reply, and thank you for taking it up with the team - I'm aware of the unified concept, and appreciate it - it certainly makes sense to be able to rely on the same behaviour, regardless of language or SDK used.

Thank you for the check-at-startup suggestion - it makes sense, of course, but alas it isn't quite automatable in the sense that you can automatically verify whether sentry received the message or not... 

I suppose a custom transport is indeed the most viable solution at the moment. Looking forward to hearing about the team's decision about this feature request!

**Comment by rhcarvalho** [comment]
> We're in a corporate setting with self-signed certificate for the sentry server. Now I know that you can set the TLS options to not enfore certificate verification, but I prefer to just use http instead.

@valentin-krasontovitsch instead of forgoing encrypted communication because of a self-signed certificate, have a look at [ClientOptions](https://godoc.org/github.com/getsentry/sentry-go#ClientOptions), you can pass in custom certificates with `CaCerts` or `HTTPTransport` or `HTTPClient`.

**Comment by rhcarvalho** [comment]
@valentin-krasontovitsch as I promised, here are some thoughts from a team discussion we had today:

-  we cannot afford to crash programs when the server is unavailable (could be a temporary network condition, etc)
- the SDKs should not send events on their own because of billing
- DSN format may be syntactically valid but still be wrong -- e.g. truncated project id. Notice that even if an event was successfully sent it could as well have been sent to a wrong project.

---

@valentin-krasontovitsch I'll need your help :sos: to shape this conversation into a more clear **feature request** that we could implement (or reject, e.g. if it doesn't make sense in the scope of the SDK). Your feedback is highly valued!

Summary of potential problems and solutions:

## HTTPS x HTTP x self-signed certs

Possible solutions:
- provide custom CaCerts as described in https://github.com/getsentry/sentry-go/issues/91#issuecomment-562352679.
- if you absolutely need to use HTTP and still often copy-pastes a DSN with HTTPS scheme, make your program validate that the DSN scheme is HTTP on startup.

## Capturing Transport errors

It can be currently done by implementing the Transport interface.
Not great, I admit, but possible.

## Crashing the app when errors are not reported to Sentry

> For us, errors in apps don't occur that often, but when they occur, it's really important that we know about them. If sentry doesn't get a message about the exception though, then we don't (get notified by email), and that's so bad, that we would like to escalate those cases. We want our app to die if it doesn't manage to report to sentry - because that's another way of triggering a message to us (the developers)...

For many programs crashing is not be the best alternative. If this is what you absolutely need your transport to do, then again the answer is a custom Transport implementation.

---

Given the points above, did I miss something?
Is there anything missing in `sentry-go` that prevents you from using it at the moment?
What would make it better or easier to handle your use case?

Thanks!

**Comment by valentin-krasontovitsch** [comment]
hi @rhcarvalho, sorry about my delayed response, and happy new year!

your comments on including your own ca certificate and validating the dsn are very valid, of course.

> we cannot afford to crash programs when the server is unavailable (could be a temporary network condition, etc)

I agree that such a behavior is probably not a good default - nevertheless having the option would be nice.

And, of course, you're right, we can just implement our own transport - and you didn't miss anything : )

What I would love to see is such a transport becoming part of the sentry library, giving users (like me) the possibility of choosing and using it. I can try to implement this, no problem, the question is mainly if this is something that you would accept if it came in as a merge request.

To get more concrete - maybe it could be an `CheckingHTTPTransport` that implements the `Transport` interface, and comes with a function 

```
func NewCheckingHTTPTransport (chan error) *CheckingHTTPTransport
```

which receives an error channel. That transport could be implemented just as HTTPTransport, except for the fact that it sends the error on the given channel in the worker if an error occurs (cf. [the corresponding worker line](https://github.com/getsentry/sentry-go/blob/c1669d9842e390a915c45b1c61d718e9b4552411/transport.go#L229), or sends nil if no error occurs.

probably `nil` also needs to be sent if the event gets dropped [due to full worker queue](https://github.com/getsentry/sentry-go/blob/c1669d9842e390a915c45b1c61d718e9b4552411/transport.go#L194), but this is technical details.

That way

- we can get a hold of an error if an error occurs during sending of an event, and handle it in any way we want
- that transport would be blocking by default, as one has to listen for a (possibly nil) error

perhaps i'm missing something, i haven't thought this all the way through and haven't tried to do a proper analysis whether this workflow could work, but assuming that the technical questions work out - does this sound like an acceptable feature to include in the SDK?

Is this concrete enough? Is it understandable? I'm happy to answer questions / react to comments.

**Comment by rhcarvalho** [comment]
Hi @valentin-krasontovitsch, happy 2020 for you too!

There is room to improve the transports, and I'd like also to make it more convenient to support your use case. Thank you for iterating on it.

I'm out of office this week and replying from my mobile. I should be back with capacity to look into this in earnest next week.

I wish to try some ideas out such that I can give you a proper educated reply. In any case, wanted to thank you and wish you a peaceful weekend!

**Comment by rhcarvalho** [comment]
> [...]
> What I would love to see is such a transport becoming part of the sentry library, giving users (like me) the possibility of choosing and using it. I can try to implement this, no problem, the question is mainly if this is something that you would accept if it came in as a merge request.
> 
> To get more concrete - maybe it could be an `CheckingHTTPTransport` that implements the `Transport` interface, and comes with a function
> 
> ```
> func NewCheckingHTTPTransport (chan error) *CheckingHTTPTransport
> ```
> [...]
> That way
> 
> * we can get a hold of an error if an error occurs during sending of an event, and handle it in any way we want
> * that transport would be blocking by default, as one has to listen for a (possibly nil) error
> 
> perhaps i'm missing something, i haven't thought this all the way through and haven't tried to do a proper analysis whether this workflow could work, but assuming that the technical questions work out - does this sound like an acceptable feature to include in the SDK?
> 
> Is this concrete enough? Is it understandable? I'm happy to answer questions / react to comments.

Hi @valentin-krasontovitsch! I've given this idea a bit more thought and synced again with the team.

Long story short, we'd rather not add more transport implementations upstream.

More transports mean more documentation, more possibilities for users to understand and choose. It would also mean detouring from other language SDKs. The default transport should do a good job most of the time. When it doesn't, users can roll out their own. We need to focus our work in improving the core transports we offer and making it easier to build on top of the core functionality.


We intentionally do not support handling of transport errors in SDKs because, by design, there are no guarantees when it comes to event delivery. Even if an event leaves the SDK successfully there are many things on the way that can swallow it: hiccup in event processing, server filters, proxies, etc.

Transport errors may happen consistently when the SDK is misconfigured. At the moment, our recommendation is to manually send a test event and manually check for it in the Sentry UI.

Users may still implement their own transport, if they so desire, and are thus free to do whatever they want there.

**Comment by rhcarvalho** [comment]
@valentin-krasontovitsch if you run into trouble rolling out your transport, please do open a new issue or [reach out to us on Discord](https://github.com/getsentry/sentry-go#community). Closing this for now. Thanks and wish you have a good experience with the new Go SDK!

---

PS: Since originally you mentioned a problem connected with misconfigured SDK and trying to programmatically validate your configuration, you may try calling the [events API](https://docs.sentry.io/api/events/get-project-event-details/) to automate "looking for an event", a single test event at the start of your program. Disclaimer: not officially supported.
